### ISSUE TITLE
How to reference variables

### ISSUE THREAD

**Issue by hferentschik** [MicrosoftDocs/vsts-docs#2757]
The [variables sections](https://docs.microsoft.com/en-us/azure/devops/pipelines/yaml-schema?view=vsts&tabs=example#variables) describes how variables can be defined, but it does not show how they then can be referenced later on (neither is there a link to where variables references would be described).

Apart of knowing the syntax on how to reference variables (I have seen {{}} as well as ${}, not sure what the difference is), I am wondering whether there are restrictions on where variables can be used. When and how are they substituted?

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: 7458d5b2-1bb7-7d87-7ce6-2b0c430b0155
* Version Independent ID: edf5bbd3-a3f0-5426-e985-31290f8afd05
* Content: [YAML schema - Azure Pipelines](https://docs.microsoft.com/en-us/azure/devops/pipelines/yaml-schema?view=vsts&tabs=example)
* Content Source: [docs/pipelines/yaml-schema.md](https://github.com/MicrosoftDocs/vsts-docs/blob/master/docs/pipelines/yaml-schema.md)
* Product: **devops**
* GitHub Login: @erickson-doug
* Microsoft Alias: **macoope**

**Comment by hferentschik** [comment]
So for example a variable cannot not be used for a task version, right? Something like:

```
  - task: oc-setup@$(taskVersion)
    displayName: Setup oc
    inputs:
      openshiftService: $(connection)
```

The _connection_ variable works fine, whereas I get an error for the task version:
>   azure-pipelines.yml (Line: 47, Col: 11): Task reference must be in the format <NAME>@<VERSION>. For example MyTask@2. The following task reference format is invalid: 'oc-setup@$(taskVersion)'

**Comment by andyjlewis** [comment]
Please check out https://docs.microsoft.com/en-us/azure/devops/pipelines/process/variables and see if that helps. I've added @vtbassmatt to advise on your specific question.

**Comment by hferentschik** [comment]
> Please check out https://docs.microsoft.com/en-us/azure/devops/pipelines/process/variables and see if that helps

That helps. A cross reference from the YAML schema page to this page would be nice. However, imo it still leaves several open questions. For example, right now I am wondering about the matrix feature. Something like:

```
  strategy:
    matrix:
      A:
        connection: 'Connection_1'
      B:
        connection: 'Connection_2'    
```

I want to run a test against two different type of service connections. In the documentation it says a variable(in this case _connection_) will be available for each generated matrix build. Are these variables equivalent to variables defined in via _variables_.  I am asking since when I am using the _connection_ variable in a custom task of mine which takes the name of a service connection   as parameter, it does not seem to get substituted:

```
steps:
- task: my-task@2
  inputs:
    myService: $(connection)
```

In this case I get an an error that the service endpoint "$(connection)" does not exist.

**Comment by vtbassmatt** [comment]
Variables are available (and considered) multiple places in the expansion/execution of a pipeline. I've been working on a doc to try and explain the different contexts, but I doubt it'll cover every possible question.

In this case, the set of service connections is decided once, when building up the job graph. (No argument whether this is right, wrong, or indifferent -- I'm just explaining the current behavior.) At that point, `connection` doesn't exist as a variable, which leads to the failure you saw. When each job resulting from the `matrix` is sent down to an agent, `connection` becomes a variable with the appropriate value. Other inputs, such as strings and display names, get a chance to evaluate `$(connection)` at that point, but it's too late for a service connection input.

**Comment by hferentschik** [comment]
>  I've been working on a doc to try and explain the different contexts, but I doubt it'll cover every possible question.

Sure, I understand there will be always open questions and corner cases, but right now I feel there is a quite big hole in general to fill

> In this case, the set of service connections is decided once, when building up the job graph. (No argument whether this is right, wrong, or indifferent -- I'm just explaining the current behaviour.)

LOL, sure. I kind of understand what's happening here. Not very intuitive imo, but as you say, you are just describing the status quo. 

For the record I could in the end achieve what I wanted to do using job templates instead of a matrix job. There I got all the substitutions I needed to happen at the right time.

I think some general section regarding job graph building and how it is send to the agent (as you did in your comment) would be a valuable addition as well.
